src/runtime/qurt_hvx.cpp

```diff
@@ -52,128 +52,26 @@ WEAK void halide_qurt_hvx_unlock_as_destructor(void *user_context, void * /*obj*
-
-    // Create the prefetch descriptor for l2fetch
-    //   l2fetch(Rs,Rtt): 48 bit descriptor
-    pdir    = pdir & 0x1;           // bit  48
-    pstride = MIN(pstride, MASK16); // bits 47:32
-    pwidth  = MIN(pwidth,  MASK16); // bits 31:16
-    pheight = MIN(pheight, MASK16); // bits 15:0
-
-    uint64_t pdir64    = pdir;
-    uint64_t pstride64 = pstride;
-    uint64_t pdesc = (pdir64<<48) | (pstride64<<32) | (pwidth<<16) | pheight;
-
-    const uint32_t pbytes = pwidth * pheight;   // Bytes in prefetch descriptor
-    uint32_t tbytes = 0;                        // Total bytes prefetched
-    uint32_t tcnt = 0;                          // Total count of prefetch ops
-
-    // If iterdim extent is unity then move to next higher dimension
-    while ((iterdim < dim_count) && (extent[iterdim] == 1)) {
-        iterdim++;
-    }
-
-    if (iterdim >= dim_count) {       // No dimension remaining to iterate over
-
-        // Perform a single prefetch
-        __asm__ __volatile__ ("l2fetch(%0,%1)" : : "r"(addr), "r"(pdesc));
-        tbytes += pbytes;
-        tcnt++;
-
-    } else {    // Iterate for higher dimension boxes
-
-        // Get iteration stride and extents
-        int32_t iterstride = stride[iterdim] * elem_size;
-        int32_t iterextent = extent[iterdim];
-        iterextent = MIN(iterextent, MAX_PREFETCH);   // Limit # of prefetches
-
-        // TODO: Add support for iterating over multiple higher dimensions?
-        // TODO  Currently just iterating over one outer (non-unity) dimension
-        // TODO  since only MAX_PREFETCH prefetches can be queued anyway.
-
-        for (int32_t i = 0; i < iterextent; i++) {
-            // Range check starting address
-            if ((addr >= addr_beg) && (addr < addr_end)) {
-                // Perform prefetch
-                __asm__ __volatile__ ("l2fetch(%0,%1)" : : "r"(addr), "r"(pdesc));
-                tbytes += pbytes;
-                tcnt++;
-            }
-            addr += iterstride;
-        }
-    }
+WEAK int halide_prefetch_2d(const void *ptr, int width_bytes, int height, int stride_bytes) {
+    // Notes:
+    //  - Prefetches can be queued up to 3 deep (MAX_PREFETCH)
+    //  - If 3 are already pending, the oldest request is dropped
+    //  - USR:PFA status bit is set to indicate that prefetches are in progress
+    //  - A l2fetch with any subfield set to zero cancels all pending prefetches
+    //  - The l2fetch starting address must be in mapped memory but the range
+    //    prefetched can go into unmapped memory without raising an exception
+    const int dir = 1;
+    uint64_t desc =
+        (static_cast<uint64_t>(dir) << 48) |
+        (static_cast<uint64_t>(stride_bytes) << 32) |
+        (static_cast<uint64_t>(width_bytes) << 16) |
+        (static_cast<uint64_t>(height) << 0);
+    __asm__ __volatile__ ("l2fetch(%0,%1)" : : "r"(ptr), "r"(desc));
+    return 0;
+}
 
-    // TODO: Opt: Return the number of prefetch instructions (tcnt) issued?
```

Comment:
I removed these TODOs because:

- We can't track prefetch counts/bytes across prefetch calls, so we'd be missing most of the cases where this went over anyways.
- Overhead here is significant.
- If we do the work to track it and stop prefetching, that's not really any worse than just going ahead with too many prefetches. All we've done in that case is at best, prefetched the wrong data and thrashed the cache anyways.